### PR TITLE
[WIP] Fix Issue 16992 - documentation does not appear for static methods

### PR DESCRIPTION
**Pull request by RazvanN7**
Even though the real bug is actually in dmd/src/ddmd/doc.d causing static inner methods to not have examples displayed, I think that having a static method with the same name in multiple structs and duplicated documentation for each one of them is bad practice. My take on this is to have a templated function which calls the appropriate static method. That way all the overloads are callable from a single public method and it is easier to publish it to users (plus you have no documentation duplication).

I implemented only fromISOString, but I plan on doing it also for fromISOExtString and for fromSimpleString.

### CODE REVIEW

**Comment by andralex** [comment]
cc @jmdavis

File: std/datetime/date.d [hunk 1]
```diff
@@ -95,6 +95,232 @@ alias AllowDayOverflow = Flag!"allowDayOverflow";
...
+
+    Throws:
+        $(REF DateTimeException,std,datetime,date) if the given string is
+        not in the ISO format or if the resulting $(LREF Date), $(LREF DateTime)
+        or $(LREF TimeOfDay) would not be valid.
+  +/
+T fromISOString(T, S)(in S isostring)
+    if (isSomeString!S)
+{
+    assert(__traits(compiles, T.fromISOString(isostring)));
```
Comment (wilzbach) [inline std/datetime/date.d]: Shouldn't this be part of the constraints?

**Comment by wilzbach** [comment]
FYI: there's also https://github.com/dlang/phobos/pull/5448 (a bit related towards adding examples to the doc)

**Comment by quickfur** [comment]
Shouldn't we be working on fixing dmd instead of hacking Phobos to make docs appear?

**Comment by andralex** [comment]
@RazvanN7 As a rule of thumb we avoid adding one-liners to Phobos unless they are of overwhelming usefulness. This PR seems to break it and I think it wouldn't be easy to construct the argument that `fromISOString` is needed with high frequency.

In addition, the notation `fromISOString!Xyz(s)` is not even shorter or more convenient than `Xyz.fromISOString` because `Xyz` cannot be deduced.

This PR could be converted to one that uses `/// Ditto` blocks to unify documentation for all overloads of `fromISOString`. It does not fix the ddoc bug which shall remain open. Thanks!

**Comment by jmdavis** [comment]
I am completely against this and think that it makes the code worse. As it stands, each of these functions is tied with the type that it goes with, just like the format is tied to the function that it goes with - much as they're all part of the same spec, they're _not_ the same format, because they don't represent the same parts of the date/time and thus the documentation should be separate as well. And Ading a free function to call them is a pointless, extra layer. Also, to call this, if you're importing the individual modules instead of all of std.datetime at once,  you'd be forced to import std.datetime.date to use this wrapper with `SysTime`, even though you can do basic stuff with a `SysTime` without ever importing std.datetime.date. I don't see how this provides any value beyond working around a dmd bug that needs to be fixed.

As far as I'm concerned, dmd needs to be fixed, and this PR should be closed.

**Comment by jmdavis** [comment]
> This PR could be converted to one that uses /// Ditto blocks to unify documentation for all overloads of fromISOString.

That doesn't even make sense, because these are functions on completely separate types, not overloads. The types follow the convention of using the same function names so that they're consistent and can easily be used in generic code where appropriate, but each of these functions expects different input (because they represent different parts of the date and/or time, even if it's from the same standard), and IMHO, it really wouldn't make sense to merge their documentation even if it were possible - which it isn't without creating a free function like this that simply calls the other functions.

The more I think about this PR, the less it makes sense to me. It's a one line wrapper around functions that can already be called in generic code if you want to. All this does is rearrange the syntax of how you call them for arguably no benefit. It also happens to work around the compiler bug that makes it so that the ddoc-ed examples don't show up in the docs, but fixing that bug would fix the issue and I really don't think that it makes any sense to rearrange things like this just to get the examples in the docs.

**Comment by jmdavis** [comment]
Okay. I'm sorry if this came across as harsh. Work to improve Phobos is much appreciated, and I certainly don't want to come across as mean or rude. But this PR just works around a compiler bug and really doesn't fix anything in std.datetime. I don't see any problem with using static functions in the way that std.datetime does with these types, and I don't see why having a wrapper function would help anything. It's just rearranging syntax and is removing documentation in the process that I think is valuable to have separately, because the types are separate, and what they expect is not the same, even if it is related enough to give the functions the same name.

**Comment by andralex** [comment]
@jmdavis oh I didn't notice the functions are not overloads. Thx for reviewing. @RazvanN7 back to the drawing board!

**Comment by RazvanN7** [comment]
@jmdavis I wasn't entirely sure either so thank you for the comprehensive explanations.

**Pull request closed without merging**